### PR TITLE
Fix all the GraphQL search/filtering issues

### LINKED ISSUE #167: Unable to query ip_addresses by parent in GraphQL
### Environment
* Python version: 3.8
* Nautobot version: 1.0.0b1 (sandbox)

### Steps to Reproduce
1. Try to query some ip addresses in GraphQL and filter by the parent prefix
```
{
  ip_addresses(parent: "10.0.65.0/24") {
    address
  }
}
```

### Expected Behavior
It should return the same data as the REST API for the same query,
In the sandbox environment the REST API returns 1 ip address ( `10.0.65.0/32` ) for this query

### Observed Behavior
Nothing is returned currently

### LINKED ISSUE #232: Graphql custom field filter query does not work
### Environment
* Python version: Sandbox
* Nautobot version: Sandbox

### Steps to Reproduce
1. Create a custom field and add some data to one object.
2.  Create a query in graphql to filter on that custom field
```
{
  devices {
    name
    custom_field_data(cf_test:"blah")
  }
}
```

### Expected Behavior
Should filter on the data.

### Observed Behavior
Returns an error

```
{
  "errors": [
    {
      "message": "Unknown argument \"cf_test\" on field \"custom_field_data\" of type \"DeviceType\".",
      "locations": [
        {
          "line": 4,
          "column": 23
        }
      ]
    }
  ]
}
```

### LINKED ISSUE #233: GraphQL boolean filters are borked
### Environment
* Python version: 3
* Nautobot version: develop

### Steps to Reproduce
1. 
```json
{
  devices(has_primary_ip:"False") {
    name
  }
}
```
2. 
```json
{
  "errors": [
    {
      "message": "unhashable type: 'list'",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "devices"
      ]
    }
  ],
  "data": {
    "devices": null
  }
}```

### Expected Behavior
A list of devices

### Observed Behavior
An error

### LINKED ISSUE #247: GraphQL : Add custom fields as filter arguments
### Environment
* Python version: 3.7
* Nautobot version: 1.0.0b2

### Proposed Functionality
Add support for custom fields in filter argument for all graphql types
Currently it's possible to define some arguments when querying a model, at the top level, but it's not possible to query by custom fields. 

if there is a custom field `type` defined under `Site`  model, it should be possible to search for cf_type
```graphql
{
  sites(cf_type: "ams") {
	id
        name
}
```
### Use Case
Similar to the API, it would be good to be able to search by custom fields with `cf_<custom_field_name>`

### Database Changes
No

### External Dependencies
No

### PR DESCRIPTION
**Pull request by jathanism**
### Fixes

This fixes several issues related to virtual or computed fields such as `IPAddress.parent` filter field, or `IPAddress.address` model field. Additionally fixes the correct type detection such as correctly detecting `Device.has_primary_ip` (filter field) as a boolean field.

- Fixes #167 
- Closes #225 (Replaces it based on the work by @carbonarok)
- Fixes #232
- Fixes #233
- Fixes #247 

#### GraphQL
- Add discrete GraphQL types in `nautobot.ipam.graphql.types` to account for the "virtual" (computed) model fields on these objects, specifically: `Aggregate.prefix`, `IPAddress.address`, `Prefix.prefix` (each of which have their corresponding `Type` object)
- Asserted that these types are registered in `nautobot.core.graphql.schema`
- Refactored `nautobot.core.graphql.generators.generate_list_search_parameters()` to call `graphene_django.filter.utils.get_filtering_args_from_filterset()` to generate the search filter params instead of doing it ourselves.
  - It patches the filterset to overload filter fields from `filterset.filters` to `filterset.base_filters` so GraphQL schema generation for list resolver query parameters detection/generation will correctly detect the custom fields and type them in the GQL schema.
- Refactored `nautobot.core.graphql.generators.generate_list_resolver()` to just pass through kwargs  since the search params are being correctly generated by `get_filtering_args_from_filterset()`.
- Register `VarbinaryIPField` as a string type in GraphQL
  - Also update tests for new schema generation (`String!` is now `[ID]!` in this test case)

#### IPAM models

Update network field types to behave as strings for humans but as varbinary for databases. This is necessary to have the fields emit correctly from Nautobot ORM without having to type cast them again for GraphQL. Less overall code to manage.

- Updated `nautobot.ipam.fields.IPNetworkFormField` to have `from_db_value`, `to_python`, and `get_db_prep_value` methods to facilitate bi-directional translation of `IPNetwork` <=> `varbinary` (packed binary int). All translation is now automatically handled by the field itself.
- Updated IPAM network models to cast objects to `str` vs. `bytes` on repr/deconstruct
- Updated IPAM network querysets to just pass objects through and not cast to `bytes`
- Updated `IPAddress` UI view to not cast `host` field to bytes and just pass it through to queryset

### Screenshots

Filtering on devices with `has_primary_ip` working as a boolean:
Cursor_and_GraphiQL_-_Nautobot
n

Filtering by interfaces `type` (mapped to `type_` due to internal GQL conflict fixed in Graphene 3.x):
GraphiQL_-_Nautobot

Single filter on devices by `site`:
GraphiQL_-_Nautobot

Multiple filter on devices by `site`:
GraphiQL_-_Nautobot

Filtering by custom fields and also included a custom field in a query schema:
GraphiQL_-_Nautobot

### CODE REVIEW

**Review by dgarros — APPROVED** [review]
Looks really good, great find with `get_filtering_args_from_filterset`
anything in `graphene_django.filter.utils` that we could leverage ?

**Comment by dgjustice** [comment]
I pulled down this branch to mess around a bit.  `has_primary_ip` is working under devices, but other API filters are not.  For example, `mgmt_only` under interfaces.  You can successfully query it as a field, but not a filter.

```
{
  devices {
    name
    interfaces(mgmt_only:true) {
      name
      ip_addresses {
        address
      }
    }
  }
}
```

**Comment by jathanism** [comment]
> I pulled down this branch to mess around a bit. `has_primary_ip` is working under devices, but other API filters are not. For example, `mgmt_only` under interfaces. You can successfully query it as a field, but not a filter.
> 
> ```
> {
>   devices {
>     name
>     interfaces(mgmt_only:true) {
>       name
>       ip_addresses {
>         address
>       }
>     }
>   }
> }
> ```

Thank you for checking on this. Haven't gotten to nested filters yet since those don't appear to be treated the same for whatever reason. Also once we get that working, need to add some tests.

**Comment by dgarros** [comment]
@dgjustice I've created a new issue this morning to track this feature https://github.com/nautobot/nautobot/issues/248

**Review by dgarros — APPROVED** [review]
GraphQL code looks good, not sure that I completely understand the changes to `VarbinaryIPField` but I trust you
With more time it would have been good to add more use unit tests for GraphQL to validate your fixes, maybe we can open an issue to track that an get someone else to help with it

**Comment by jathanism** [comment]
> GraphQL code looks good, not sure that I completely understand the changes to `VarbinaryIPField` but I trust you
> 
> With more time it would have been good to add more use unit tests for GraphQL to validate your fixes, maybe we can open an issue to track that an get someone else to help with it

The changes I made to the network fields to emit strings for humans and packed int for databases will greatly simplify exports. I solved the problem centrally to obviate the need for callers (GraphQL in this case) to have to care about doing the binary unpacking. If I didn't do it here, we would have to duplicate the binary conversions in the GraphQL code.

You can see the result of it when you attempt to query on one of the fields (like `Prefix.network` in GraphQL in `develop`. It shows as a byte array (memory view) instead of say "10.0.0.0".

**Comment by jathanism** [comment]
@dgarros 

Before `VarbinaryIPField` changes:
GraphiQL_-_Nautobot

After: 
Cursor_and_GraphiQL_-_Nautobot

Which also means working with these from the ORM will be more intuitive (even though these fields aren't part of the writable public interface).

File: nautobot/core/graphql/generators.py [hunk 1]
```diff
@@ -77,20 +79,22 @@ def generate_schema_type(app_name: str, model: object) -> DjangoObjectType:
...
-        Class DeviceType(DjangoObjectType):
+        class DeviceType(DjangoObjectType):
             Meta:
                 model = Device
                 fields = ["__all__"]
 
-        if a FilterSet exist for this model at '<app_name>.filters.<ModelName>FilterSet'
-        The filterset will be store in filterset_class as follow
+        If a FilterSet exists for this model at
+        '<app_name>.filters.<ModelName>FilterSet' the filterset will be store in
```
Comment (glennmatthews) [inline nautobot/core/graphql/generators.py:88]:
```suggestion
        '<app_name>.filters.<ModelName>FilterSet' the filterset will be stored in
```

File: nautobot/core/graphql/generators.py [hunk 2]
```diff
@@ -108,15 +112,23 @@ def generate_list_search_parameters(schema_type):
...
+        search_params = get_filtering_args_from_filterset(
+            filterset,
+            schema_type,
+        )
+
+    # Hack to swap `type` fields to `type_` since they will conflict with
+    # `graphene.types.fields.Field.type`.
+    # FIXME(jathan): Once we upgrade to Graphene 3.x, remove this.
+    if "type" in search_params:
+        search_params["type_"] = search_params.pop("type")
```
Comment (glennmatthews) [inline nautobot/core/graphql/generators.py:131]: Do I understand correctly that Graphene 3.x reverses this (i.e., `type` is a valid variable name but `type_` will conflict)? If so, should we pick a name (such as `_type` or `type__`) that will work as-is regardless of the Graphene version? Trying to reduce our need for breaking API changes in the future. :-)

Comment (jathanism) [reply nautobot/core/graphql/generators.py:131]: It's both. It ends up being stored as `self._type` on the instance. Both of them just unpack kwargs (as `**extra_args`, and stores them on `self.args`.

- Graphene v2.1.8 (what we are currently using) [uses `type` as the argument](https://github.com/graphql-python/graphene/blob/v2.1.8/graphene/types/field.py#L65)
- Graphene v3.x uses [`type_` as the argument](https://github.com/graphql-python/graphene/blob/v3.0.0b7/graphene/types/field.py#L68)

The other option is to just upgrade to Graphene 3.0.0b7+ now and cut out the middleman. I attempted to do that, but opened up a rabbit hole on an API change in `graphql-core` that no longer requires GraphQL "backends" to be defined. (We're using `from graphql import get_default_backend`) in [`nautobot.core.api.views.GraphQLDRFAPIView`](https://github.com/nautobot/nautobot/blob/develop/nautobot/core/api/views.py#L360)). It wasn't a quick win, but might be worth it just to future proof so we could just roll forward with `type` unhindered.

Comment (jathanism) [reply nautobot/core/graphql/generators.py:131]: I revisited this again to try to upgrade to Graphene 3.x and failed, but I did validate that renaming to `_type` will be safe for both 2.x and 3.x, so I will swap that.

File: nautobot/ipam/tests/test_models.py [hunk 3]
```diff
@@ -1,12 +1,139 @@
...
+        Prefix.objects.create
+
+    def test_update(self):
+        """Test network object update."""
+
+    def test_get(self):
+        """Test retrieve of single network object."""
+
+    def test_filter(self):
+        """Test filtering list of network objects."""
```
Comment (jathanism) [inline nautobot/ipam/tests/test_models.py:134]: LOL. This wasn't supposed to be committed. But hey tests are passing! 👀

**Review by glennmatthews — APPROVED** [review]
Looks good to me. Thanks for adding test cases!

**Pull request merged**